### PR TITLE
[TEST MERGE] - Code Standardisation - DNA Procedures & New DNA Type

### PR DESCRIPTION
**Pull request by Kiyahitayika**
## What Does This PR Do
What it says in the title.
This PR standardizes the code for the DNA procs. The procs for SE and the procs for UI have always been copy/pastas of each other - some with small differences, others exactly the same, some with the _comments above them_ exactly the same (one comment for an SE proc said "don't use this, use this UI proc", revealing the copy/pasta truth!).
These procs have been combined now.

Also adds new TYPE of DNA (RP) - this is easy to remove if other servers want to use this code! Additionally, it isn't fully implemented. 

## Why It's Good For The Game

1. Code Standardization = Good
2. Makes it easy to add new types of DNA. 
3. Relevant to Issue: https://github.com/ScorpioStation/ScorpioStation/issues/274 Relevant for Steps 2 & 3 of PR: https://github.com/ScorpioStation/ScorpioStation/issues/274
4. 27 procs reduced to **15 procs**
5. Replaced code block used repeatedly with proc called `ValidCheck()` 

## New Proc Syntax
**As `defer` is an optional argument, it was kept at the end.**
- UpdateDNA(dna_type) ---> *Accepts DNA_ALL*
- ResetDNA(dna_type, defer = FALSE) ---> *Accepts DNA_ALL*
- ResetDNAFrom(mob/living/carbon/human/character, dna_type) --> *Previously there was only ResetUIFrom(), so anything other than DNA_UI returns.*
=
- SetDNAState(block, on, dna_type, defer = FALSE)
- GetDNAState(block, dna_type)
- SetDNATriState(block, value, dna_type, defer = FALSE)
- GetDNATriState(block, dna_type)
=
- SetDNABlock(block, newBlock, dna_type, defer = FALSE)
- GetDNABlock(block, dna_type)
- SetDNASubBlock(block, subBlock, newSubBlock, dna_type, defer = FALSE)
- GetDNASubBlock(block, subBlock, dna_type)
=
- SetDNAValue(block, value, dna_type, defer = FALSE)
- GetDNAValue(block, dna_type)
- SetDNAValueRange(block, value, maxvalue, dna_type, defer = FALSE)
- GetDNAValueRange(block, maxvalue, dna_type)

## Changelog
:cl:
add: New DNA Type - DNA.RP to be used for Roleplaying Preferences (RP)
tweak: Code for DNA
/:cl:

### CODE REVIEW

File: code/game/gamemodes/miniantags/abduction/gland.dm [hunk 1]
```diff
@@ -280,13 +280,13 @@
...
 		owner.gene_stability += GENE_INSTABILITY_MODERATE // give them this gene for free
-		owner.dna.SetSEState(GLOB.shockimmunityblock, TRUE)
+		owner.dna.SetDNAState(GLOB.shockimmunityblock, DNA_SE, TRUE)
 		genemutcheck(owner, GLOB.shockimmunityblock,  null, MUTCHK_FORCED)
 
 /obj/item/organ/internal/heart/gland/electric/remove(mob/living/carbon/M, special = 0)
 	if(ishuman(owner))
 		owner.gene_stability -= GENE_INSTABILITY_MODERATE // but return it to normal once it's removed
-		owner.dna.SetSEState(GLOB.shockimmunityblock, FALSE)
+		owner.dna.SetDNAState(GLOB.shockimmunityblock, FALSE, DNA_SE)
```
Comment (blinkdog) [inline code/game/gamemodes/miniantags/abduction/gland.dm:289]: There are two different orderings here.
I think the second one is right?

File: code/game/jobs/job/support.dm [hunk 2]
```diff
@@ -370,11 +370,11 @@
 		var/obj/item/organ/internal/cyberimp/brain/clown_voice/implant = new
 		implant.insert(H)
 
-	H.dna.SetSEState(GLOB.clumsyblock, TRUE)
+	H.dna.SetDNAState(GLOB.clumsyblock, DNA_SE, TRUE)
 	genemutcheck(H, GLOB.clumsyblock, null, MUTCHK_FORCED)
 	H.dna.default_blocks.Add(GLOB.clumsyblock)
 	if(!ismachineperson(H))
-		H.dna.SetSEState(GLOB.comicblock, TRUE)
+		H.dna.SetDNAState(GLOB.comicblock, DNA_SE, TRUE)
```
Comment (blinkdog) [inline code/game/jobs/job/support.dm:377]: I think the orderings here need to be `GLOB.comicblock, TRUE, DNA_SE` ?

File: code/modules/admin/topic.dm [hunk 3]
```diff
@@ -2015,15 +2015,15 @@
...
+				H.dna.SetDNAState(GLOB.regenerateblock, DNA_SE, TRUE)
 				genemutcheck(H, GLOB.regenerateblock,  null, MUTCHK_FORCED)
 				H.update_mutations()
 				H.gene_stability = 100
 				logmsg = "permanent regeneration."
 			if("Super Powers")
 				var/list/default_genes = list(GLOB.regenerateblock, GLOB.breathlessblock, GLOB.coldblock)
 				for(var/gene in default_genes)
-					H.dna.SetSEState(gene, 1)
+					H.dna.SetDNAState(gene, DNA_SE, TRUE)
```
Comment (blinkdog) [inline code/modules/admin/topic.dm:2026]: I think the ordering for these two here needs to be `block, TRUE, DNA_SE` ?

File: code/modules/antagonists/wishgranter/wishgranter.dm [hunk 4]
```diff
@@ -22,55 +22,55 @@
...
-	H.dna.SetSEState(GLOB.eatblock, TRUE)
+	H.dna.SetDNAState(GLOB.eatblock, DNA_SE, TRUE)
 	genemutcheck(H, GLOB.eatblock, null, MUTCHK_FORCED)
 
-	H.dna.SetSEState(GLOB.jumpblock, TRUE)
+	H.dna.SetDNAState(GLOB.jumpblock, DNA_SE, TRUE)
 	genemutcheck(H, GLOB.jumpblock, null, MUTCHK_FORCED)
 
-	H.dna.SetSEState(GLOB.immolateblock, TRUE)
+	H.dna.SetDNAState(GLOB.immolateblock, DNA_SE, TRUE)
```
Comment (blinkdog) [inline code/modules/antagonists/wishgranter/wishgranter.dm:73]: I think all of these have `DNA_SE` and `TRUE` in the reverse order?

**Review by blinkdog — CHANGES_REQUESTED** [review]
A few things to fix up and a couple of others to consider.

Notes for myself:
I need to re-read it fresh from the top to re-check all the SetDNAState calls.
I also need to check cloning.dm:168 against the function signature for setting a range.

File: code/modules/antagonists/wishgranter/wishgranter.dm (hunk 4 above)
Comment (Kiyahitayika) [reply code/modules/antagonists/wishgranter/wishgranter.dm:73]: Huh. What happened here . . .

File: code/datums/spells/cluwne.dm [hunk 5]
```diff
@@ -27,7 +27,7 @@
 	var/obj/item/organ/internal/honktumor/cursed/tumor = new
 	tumor.insert(src)
 	mutations.Add(NERVOUS)
-	dna.SetSEState(GLOB.nervousblock, 1, 1)
+	dna.SetDNAState(GLOB.nervousblock, 1, DNA_SE, TRUE)
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/datums/spells/cluwne.dm:30]: That is a good question. I'll double check. Most likely, yes.

File: code/datums/spells/wizard.dm [hunk 6]
```diff
@@ -97,10 +97,10 @@
 
 /obj/effect/proc_holder/spell/targeted/genetic/mutate/cast(list/targets, mob/user = usr)
 	for(var/mob/living/target in targets)
-		target.dna.SetSEState(GLOB.hulkblock, 1)
+		target.dna.SetDNAState(GLOB.hulkblock, DNA_SE, TRUE)
 		genemutcheck(target, GLOB.hulkblock, null, MUTCHK_FORCED)
 		spawn(duration)
-			target.dna.SetSEState(GLOB.hulkblock, 0)
+			target.dna.SetDNAState(GLOB.hulkblock, FALSE, DNA_SE)
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/datums/spells/wizard.dm:103]: This is odd. Same as below - how did this happen? This is where I really appreciate a second set of eyes; thank you for reviewing this. 

Second one is correct.
First one, I need to check the line it replaced. There's an optional `defer` variable that comes after the `dna_type` variable, but the on or off is not optional.

`SetDNAState(block, on/off, dna_type, (optional) defer)`

File: code/game/dna/dna2.dm [hunk 7]
```diff
@@ -55,35 +54,72 @@ GLOBAL_LIST_EMPTY(bad_blocks)
 	new_dna.blood_type = blood_type
 	new_dna.real_name = real_name
 	new_dna.species = new species.type
-	for(var/b=1;b<=DNA_SE_LENGTH;b++)
+	for(var/b=1;b<=DNA_SE_LENGTH;b++) //Should DNA_RP be involved in here somewhere? Iunno!
 		new_dna.SE[b]=SE[b]
 		if(b<=DNA_UI_LENGTH)
 			new_dna.UI[b]=UI[b]
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:60]: Oh yikes! Thank you!! Thank you!!!!!

File: code/game/dna/dna2.dm [hunk 8]
```diff
@@ -102,112 +138,149 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
-		SetUIValue(block,round(value * range), defer)
+	switch(dna_type)
+		if(DNA_UI)			// Used in hair and facial styles (value being the index and maxvalue being the len of the hairstyle list)
+			if(value == 0)
+				value = 1
+			SetDNAValue(block, round(value * range), DNA_UI, defer)
+		if(DNA_SE)		// Might be used for species?
+			SetDNAValue(block, round(value * range) - rand(1, range - 1), DNA_SE)
+		if(DNA_RP)
+			SetDNAValue(block, round(value * range), DNA_RP)
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:232]: Great question. I'll need to brush up on the code before answering your question. I had the answer previously, but I need to find it again. :)

File: code/game/dna/dna2.dm [hunk 9]
```diff
@@ -102,112 +138,149 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
+			else		//Or to "off" (FALSE)?
+				val = rand(1, 2049)
+		if(DNA_SE)
+			var/list/BOUNDS=GetDNABounds(block)
+			if(on)
+				val = rand(BOUNDS[DNA_ON_LOWERBOUND], BOUNDS[DNA_ON_UPPERBOUND])
+			else
+				val = rand(1, BOUNDS[DNA_OFF_UPPERBOUND])
+		if(DNA_RP)
+			val = on		// DNA_RP is simple, okay? "on" is on and "off" is off!
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:264]: Better code is better code. I definitely appreciate the recommendation and will change it accordingly!

File: code/game/gamemodes/changeling/powers/chameleon_skin.dm [hunk 10]
```diff
@@ -11,19 +11,19 @@
...
 	if(!istype(H)) // req_human could be done in can_sting stuff.
 		return
-	if(H.dna.GetSEState(GLOB.chameleonblock))
-		H.dna.SetSEState(GLOB.chameleonblock, 0)
+	if(H.dna.GetDNAState(GLOB.chameleonblock, DNA_SE))
+		H.dna.SetDNAState(GLOB.chameleonblock, FALSE, DNA_SE)
 		genemutcheck(H, GLOB.chameleonblock, null, MUTCHK_FORCED)
 	else
-		H.dna.SetSEState(GLOB.chameleonblock, 1)
+		H.dna.SetDNAState(GLOB.chameleonblock, DNA_SE, TRUE)
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/gamemodes/changeling/powers/chameleon_skin.dm:18]: Yeah, this is weird. I must ha-- Aaah. I *definitely* did a Regex type find-and-replace to grab what I thought was the `defer` variable before changing some of the `1`s to TRUE. This is what happens when I try to repeat my work from scratch, but forget the order. >__<;

File: code/datums/spells/cluwne.dm (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/datums/spells/cluwne.dm:30]: Yes!

File: code/game/gamemodes/miniantags/abduction/gland.dm (hunk 1 above)
Comment (Kiyahitayika) [reply code/game/gamemodes/miniantags/abduction/gland.dm:289]: Yep. This was a bad find-and-replace. Looks like I did it correctly for 0/FALSE. Really appreciate this review, seriously. Thank you so much.

File: code/game/jobs/job/support.dm (hunk 2 above)
Comment (Kiyahitayika) [reply code/game/jobs/job/support.dm:377]: Nice catch!

File: code/modules/admin/topic.dm (hunk 3 above)
Comment (Kiyahitayika) [reply code/modules/admin/topic.dm:2026]: They sure do x__x;

File: code/game/dna/dna2_helpers.dm [hunk 11]
```diff
@@ -51,15 +51,15 @@
...
 		M.UpdateAppearance()
 
 	else
 		for(var/i = 1, i <= DNA_SE_LENGTH - 1, i++)
 			if(prob(prob))
-				M.dna.SetSEValue(i, rand(1, 4095), 1)
-		M.dna.UpdateSE()
+				M.dna.SetDNAValue(i, rand(1, 4095), DNA_SE, TRUE)
+		M.dna.UpdateDNA(DNA_SE)
 		domutcheck(M, null)
```
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2_helpers.dm:63]: For DNA_SE, it is intentional. The last block of SE is 55. 55 indicates the length of SE.
For UI, that's a good question.

[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2_helpers.dm:63]: I'm going to change that `else` to `if(SE)`.
I'm going to update these `for` loops with new syntax: 
`for(var/i in 1 to DNA_UI_LENGTH)`
`for(var/i in 1 to (DNA_SE_LENGTH - 1))`

[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2_helpers.dm:63]: Really good catch, thank you!

File: code/game/dna/dna2.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:232]: This is a two-part question. 

- Q1: Is `defer` used for DNA_SE or DNA_RP?
- A1: No, `defer` is not used for DNA_SE nor DNA_RP.

```
if(!defer)
	UpdateDNA(DNA_UI)
```

- Q2: Even if it is not used, does quality coding dictate that we pass the argument along?
- A2: Ultimately, the `defer` argument is called in `ResetDNA`. This was previously two procs (ResetSE and ResetUI). In ResetUI, `defer` was called upon, in ResetSE, it was not. In order to preserve the argument, we pass it down through each proc that `ResetDNA` relies upon. This is why it is passed in `SetDNAValueRange` for DNA_UI.
We *can* pass it for DNA_SE and DNA_RP as well for a matter of consistency.

[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:232]: The reason the code does not UpdateDNA when `!defer` is because if you are changing, for example, skin color for some species, you are changing _three_ DNA_UI values at once:

```
	SetDNAValueRange(DNA_UI_SKIN_R,		color2R(character.skin_colour),			255,	DNA_UI,	TRUE)
	SetDNAValueRange(DNA_UI_SKIN_G,		color2G(character.skin_colour),			255,	DNA_UI,	TRUE)
	SetDNAValueRange(DNA_UI_SKIN_B,		color2B(character.skin_colour),			255,	DNA_UI,	TRUE)
```
By ***deferring*** until all the three values are changed, the skin color for the mob sprite is updated appropriately instead of R, G, and B individually.

**Comment by Kiyahitayika** [comment]
> I also need to check cloning.dm:168 against the function signature for setting a range.

`buf.dna.SetDNAValueRange(GLOB.monkeyblock, 0xDAC, 0xFFF, DNA_SE)`
What the heck am I looking at? 👀

**Comment by Kiyahitayika** [comment]
> I need to re-read it fresh from the top to re-check all the SetDNAState calls.

Just did this as well. I two additional instances. I'm going to push a commit with all these changes. Please let me know if you see more.

File: code/game/machinery/cloning.dm [hunk 12]
```diff
@@ -165,7 +165,7 @@ GLOBAL_LIST_INIT(cloner_biomass_items, list(\
 	for(var/i=new_SE.len;i<=DNA_SE_LENGTH;i++)
 		new_SE += rand(1,1024)
 	buf.dna.SE=new_SE
-	buf.dna.SetSEValueRange(GLOB.monkeyblock,0xDAC, 0xFFF)
+	buf.dna.SetDNAValueRange(GLOB.monkeyblock, 0xDAC, 0xFFF, DNA_SE)
```
Comment (Kiyahitayika) [inline code/game/machinery/cloning.dm:168]: @blinkdog 
> I also need to check cloning.dm:168 against the function signature for setting a range.

`buf.dna.SetDNAValueRange(GLOB.monkeyblock, 0xDAC, 0xFFF, DNA_SE)`
\---------------
```
/datum/dna/proc/SetDNAValueRange(block, value, maxvalue, dna_type, defer = FALSE)
var/range = (4095 / maxvalue)`
SetDNAValue(block, round(value * range) - rand(1, range - 1), DNA_SE)
```
\---------------
0xDAC --> 3500
0xFFF --> 4095

SetDNAValueRange(GLOB.monkeyblock, 3500, 4095, DNA_SE, FALSE)
\---------------
var/range = (4095 / 4095) = 1
SetDNAValue(GLOB.monkeyblock, round(3500 * 1) - rand(1, 1 - 1), DNA_SE)
\---------------
SetDNAValue(GLOB.monkeyblock, 3500 - rand(1, 0), DNA_SE)
\---------------
SetDNAValue(GLOB.monkeyblock, 3500, DNA_SE)
\---------------

SE uses Value Ranges to randomize the numbers on the DNA Modifier machines in Genetics. It's a really cool system. It encodes using hexadecimal here.

_This_, on the other hand, is hard-coding a value-range set. I'm not sure what the purpose is. This _should_ be using `SetDNAValue`.
\---------------
```
/datum/dna/proc/SetDNAValue(block, value, dna_type, defer = FALSE)`
		if(DNA_SE)
			SE[block] = value
			dirtySE = TRUE
	if(!defer)
		UpdateDNA(dna_type)		
```
\---------------
SetDNAValue(GLOB.monkeyblock, 3500, DNA_SE)

defer --> FALSE

SE[GLOB.monkeyblock] = 3500
dirtySE = TRUE

UpdateDNA(DNA_SE)
\---------------
```
/datum/dna/proc/UpdateDNA(dna_type)
		if(DNA_SE)
			struc_enzymes = ""
			for(var/block in SE)
				struc_enzymes += EncodeDNABlock(block)
			dirtySE = FALSE
```
\---------------
struc_enzymes += EncodeDNABlock(GLOB.monkeyblock)

`/proc/EncodeDNABlock(value)
	return add_zero2(num2hex(value, 1), 3)`

add_zero2(num2hex(3500, 1), 3)
add_zero2(0xDAC, 3)
0000xDAC

struc_enzymes += 0000xDAC
\---------------

File: code/game/machinery/cloning.dm [hunk 13]
```diff
@@ -165,7 +165,7 @@ GLOBAL_LIST_INIT(cloner_biomass_items, list(\
 	for(var/i=new_SE.len;i<=DNA_SE_LENGTH;i++)
```
Comment (Kiyahitayika) [inline code/game/machinery/cloning.dm:165]: Let's also talk about the TODO comment at the top of this code - line 121. Who ever wrote this...Why?! WHY IS THIS CODE PASSING HEXADECIMAL?!

Why isn't this using the actual DNA code? What the heck is 
`buf.dna.UI = list(0x066,0x000,0x033,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0xAF0,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0x000,0x033,0x066,0x0FF,0x4DB,0x002,0x690,0x000,0x000,0x000,0x328,0x045,0x5FC,0x053,0x035,0x035,0x035)
`
I have no clue which UI values these are.

`var/list/new_SE=list(0x098,0x3E8,0x403,0x44C,0x39F,0x4B0,0x59D,0x514,0x5FC,0x578,0x5DC,0x640,0x6A4)`

What SE values are these?!

Is this it? Is THIS the problem? This spaghetti? 
```
//The return of data disks?? Just for transferring between genetics machine/cloning machine.
//TO-DO: Make the genetics machine accept them.
```
These disks aren't even used! Thanks, GOON!

Comment (Kiyahitayika) [reply code/game/machinery/cloning.dm:165]: image

Comment (Kiyahitayika) [reply code/game/machinery/cloning.dm]: Note: DISKS are used.
The `/disk/data/monkey` disk is not used.

Comment (Kiyahitayika) [reply code/game/machinery/cloning.dm]: @blinkdog I'd be happy to take a closer look at these disks to clarify the code in the future. 

TL;DR: It's passing Hexadecimal directly into the SE and UI lists. That's ... okay for what's it's doing - setting up hard-coded DNA for this disk type. That's real, that's fair. So, eeeh, I went through and made it messier:

```
	buf.dna.UI = list(
		0x066,	//DNA_UI_HAIR_R
		0x000,	//DNA_UI_HAIR_G
		0x033,	//DNA_UI_HAIR_B
		0x000,	//DNA_UI_HAIR_R2
		0x000,	//DNA_UI_HAIR_G2
		0x000,	//DNA_UI_HAIR_B2
		0x000,	//DNA_UI_BEARD_R
		0x000,	//DNA_UI_BEARD_G
		0x000,	//DNA_UI_BEARD_B
		0x000,	//DNA_UI_BEARD2_R
		0x000,	//DNA_UI_BEARD2_G
		0x000,	//DNA_UI_BEARD2_B
		0xAF0,	//DNA_UI_SKIN_TONE
		0x000,	//DNA_UI_SKIN_R
		0x000,	//DNA_UI_SKIN_G
		0x000,	//DNA_UI_SKIN_B
		0x000,	//DNA_UI_HACC_R
		0x000,	//DNA_UI_HACC_G
		0x000,	//DNA_UI_HACC_B
		0x000,	//DNA_UI_HEAD_MARK_R
		0x000,	//DNA_UI_HEAD_MARK_G
		0x000,	//DNA_UI_HEAD_MARK_B
		0x033,	//DNA_UI_BODY_MARK_R
		0x066,	//DNA_UI_BODY_MARK_G
		0x0FF,	//DNA_UI_BODY_MARK_B
		0x4DB,	//DNA_UI_TAIL_MARK_R
		0x002,	//DNA_UI_TAIL_MARK_B
		0x690,	//DNA_UI_TAIL_MARK_G
		0x000,	//DNA_UI_EYES_R
		0x000,	//DNA_UI_EYES_G
		0x000,	//DNA_UI_EYES_B
		0x328,	//DNA_UI_GENDER
		0x045,	//DNA_UI_BEARD_STYLE
		0x5FC,	//DNA_UI_HAIR_STYLE
		0x053,	//DNA_UI_HACC_STYLE
		0x035,	//DNA_UI_HEAD_MARK_STYLE
		0x035,	//DNA_UI_BODY_MARK_STYLE
		0x035	//DNA_UI_TAIL_MARK_STYLE
	)
```

Comment (Kiyahitayika) [reply code/game/machinery/cloning.dm]: I feel like I messed something up. Why would there be a tail mark style?

File: code/game/dna/dna2.dm [hunk 14]
```diff
@@ -102,112 +140,152 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
+		if(DNA_UI)
+			return UI[block]
+		if(DNA_SE)
+			return SE[block]
+		if(DNA_RP)
+			return RP[block]
+
+// Set a DNA block's value, given a value and a max possible value
+/datum/dna/proc/SetDNAValueRange(block, value, maxvalue, dna_type, defer = FALSE)
+	if(!ValidCheck(block, dna_type) || !value)
```
Comment (blinkdog) [inline code/game/dna/dna2.dm:222]: Does the second conditional (`!value`) prevent DNA blocks from being set to a value of `0` ?

Maybe better to check if `value < 0 || value > maxvalue` ?

File: code/game/dna/dna2.dm [hunk 15]
```diff
@@ -55,35 +54,74 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
+			for(var/i in 1 to DNA_UI_LENGTH)
+				if(i == DNA_UI_SKIN_TONE)
+					SetDNAValueRange(DNA_UI_SKIN_TONE, rand(1, 220), 220, DNA_UI, TRUE)	// Otherwise, it gets screwed up
+				else
+					UI[i] = rand(0, 4095)
+			if(!defer)
+				UpdateDNA(DNA_UI)
+		if(DNA_SE)	// "Zeroes out" all of the SE Blocks
+			for(var/i in 1 to (DNA_SE_LENGTH - 1))
+				SetDNAValue(i, rand(1, 1024), TRUE, DNA_SE)
```
Comment (blinkdog) [inline code/game/dna/dna2.dm:84]: Ordering here looks like it should be:
`SetDNAValue(i, rand(1, 1024), DNA_SE, TRUE)`

File: code/game/dna/dna2.dm [hunk 16]
```diff
@@ -55,35 +54,74 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
+			UpdateDNA(DNA_SE)
+		if(DNA_RP)
+			for(var/i in 1 to DNA_RP_LENGTH)
+				SetDNAValue(i, rand(1, 1024), TRUE, DNA_RP)
```
Comment (blinkdog) [inline code/game/dna/dna2.dm:88]: Ordering here looks like it should be:
`SetDNAValue(i, rand(1, 1024), DNA_RP, TRUE)`

File: code/game/machinery/cloning.dm [hunk 17]
```diff
@@ -140,32 +140,32 @@ GLOBAL_LIST_INIT(cloner_biomass_items, list(\
...
+	buf.types = DNA2_BUF_SE
 	var/list/new_SE=list(0x098,0x3E8,0x403,0x44C,0x39F,0x4B0,0x59D,0x514,0x5FC,0x578,0x5DC,0x640,0x6A4)
-	for(var/i=new_SE.len;i<=DNA_SE_LENGTH;i++)
-		new_SE += rand(1,1024)
-	buf.dna.SE=new_SE
-	buf.dna.SetSEValueRange(GLOB.monkeyblock,0xDAC, 0xFFF)
+	for(var/i in length(new_SE) to DNA_SE_LENGTH)
+		new_SE += rand(1, 1024)
+	buf.dna.SE = new_SE
+	buf.dna.SetDNAValue(GLOB.monkeyblock, 3500, DNA_SE)
```
Comment (blinkdog) [inline code/game/machinery/cloning.dm:168]: I think `0xDAC` is more expressive than `3500` here, as "DAC" matches what a geneticist needs in-game to activate a block.

Consider changing it back?

**Review by blinkdog — CHANGES_REQUESTED** [review]
A few more things to fix up, but getting pretty close!

File: code/game/machinery/cloning.dm (hunk 17 above)
Comment (Kiyahitayika) [reply code/game/machinery/cloning.dm:168]: Absolutely! 👍

File: code/game/dna/dna2.dm (hunk 16 above)
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:88]: Fascinating.  I am impressed by myself. How did I miss these things?
I, myself, was looking at this the other day and was thinking, "Huh. That order looks wrong" and, yet, I couldn't see it. Thank you!

File: code/game/dna/dna2.dm (hunk 15 above)
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:84]: The ordering, in fact, should indeed be what you suggest!

Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:84]: Wait, hold up. Why am I trying to pass a defer = TRUE to`DNA_RP` and `DNA_SE`?

```
/datum/dna/proc/ResetSE()
	for(var/i = 1, i <= DNA_SE_LENGTH, i++)
		SetSEValue(i, rand(1, 1024), 1)
	UpdateSE()
```

File: code/game/dna/dna2.dm [hunk 18]
```diff
@@ -102,112 +140,152 @@ GLOBAL_LIST_EMPTY(bad_blocks)
...
+			dirtyUI = TRUE
+		if(DNA_SE)
+			SE[block] = value
+			dirtySE = TRUE
+		if(DNA_RP)
+			RP[block] = value
+			dirtyRP = TRUE
 	if(!defer)
-		UpdateUI()
+		UpdateDNA(dna_type)
```
Comment (Kiyahitayika) [inline code/game/dna/dna2.dm:207]: @blinkdog Calling myself out from earlier. You were correct on passing `defer`. Yikes.

File: code/game/dna/dna2.dm (hunk 14 above)
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:222]: This proc was originally `SetUIValueRange()`.

I combined it with `SetSEValueRange()` which did not have the same `if(value == 0)` check, so that check only happens if the `dna_type` is `DNA_UI`.

```
/datum/dna/proc/SetSEValueRange(block, value, maxvalue)
	if(block <= 0)
		return
	ASSERT(maxvalue <= 4095)
	var/range = round(4095 / maxvalue)
	if(value)
		SetSEValue(block, value * range - rand(1, range - 1))
```

A-And, hang on . . . hold up . . . 

image

I do not currently see a use case for using `SetSEValueRange()`, now `SetDNAValueRange(DNA_SE)` and this code supports the fact that using it does not make much sense.  The scope of this PR was mainly to combine the procs which involved changing syntax. This review has brought up some really great questions. At this point, I think it might be best to code supporting setting value ranges for SE (commented out?). But, let's move that checker up some levels.

File: code/game/dna/dna2.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kiyahitayika) [reply code/game/dna/dna2.dm:234]: @Kiyahitayika Hey smarty-pants, I think you're wrong.
